### PR TITLE
fix(lib/runtime): allocator out of space and wasm out of bounds memory access

### LINKED ISSUE #3774: (lib/runtime): fail to execute block `17088245` allocator out of space
```
2024-02-24T03:55:37Z INFO     imported block 0x2aad31837b6cdf8ce5964cca86b3a39b67ff872aa5eb6c954217872f6683d63d and stored state trie with root 0xe40e9900f63d1539014ce5cc4fa0e4932ae4ed1154c31cd77714d0fdf2d4e8b3	pkg=core
RUNTIME HASH: 0xd1d708433e5d7bcb9fc30a3de67952c1611fa58cd1980146ab07aa438116d352
2024-02-24T03:55:37Z ERROR    block data processing for block with hash 0x0feb9c8bb9018bbcf20b467e15636de02397de00c66233c5b4a06f5be3a2f558 failed: processing block data with header and body: handling block: failed to execute block 17088245: running runtime function: bumping: allocator out of space: current pages 65535 greater than max wasm pages 65535 (recovered by wazero)
wasm stack trace:
	env.ext_allocator_malloc_version_1(i32) i32
	._ZN5sp_io9allocator26extern_host_function_impls6malloc17h175ebffadfcb2d44E(i32) i32
	._ZN5alloc7raw_vec11finish_grow17h97bec91bef32a5ebE.llvm.1609301459894010017(i32,i32,i32,i32)
	._ZN5alloc7raw_vec19RawVec$LT$T$C$A$GT$16reserve_for_push17h1ed9c357b734a946E(i32,i32)
	._ZN18parity_scale_codec5codec19decode_vec_with_len17h15e2f12c18cd243fE(i32,i32,i32)
	._ZN13frame_support7storage8unhashed3get17hb6cef7316d8fee28E(i32,i32,i32)
	._ZN353_$LT$pallet_staking..ExposureOf$LT$T$GT$$u20$as$u20$sp_runtime..traits..Convert$LT$$LT$T$u20$as$u20$frame_system..pallet..Config$GT$..AccountId$C$core..option..Option$LT$pallet_staking..Exposure$LT$$LT$T$u20$as$u20$frame_system..pallet..Config$GT$..AccountId$C$$LT$T$u20$as$u20$pallet_staking..pallet..pallet..Config$GT$..CurrencyBalance$GT$$GT$$GT$$GT$7convert17h469fc883609fa8e1E(i32,i32)
	._ZN5alloc3vec16in_place_collect108_$LT$impl$u20$alloc..vec..spec_from_iter..SpecFromIter$LT$T$C$I$GT$$u20$for$u20$alloc..vec..Vec$LT$T$GT$$GT$9from_iter17h34924d7a99a1d609E(i32,i32)
	._ZN14pallet_session10historical7onchain39store_session_validator_set_to_offchain17h0e99a24533159ab9E(i32)
	._ZN893_$LT$$LP$TupleElement0$C$TupleElement1$C$TupleElement2$C$TupleElement3$C$TupleElement4$C$TupleElement5$C$TupleElement6$C$TupleElement7$C$TupleElement8$C$TupleElement9$C$TupleElement10$C$TupleElement11$C$TupleElement12$C$TupleElement13$C$TupleElement14$C$TupleElement15$C$TupleElement16$C$TupleElement17$C$TupleElement18$C$TupleElement19$C$TupleElement20$C$TupleElement21$C$TupleElement22$C$TupleElement23$C$TupleElement24$C$TupleElement25$C$TupleElement26$C$TupleElement27$C$TupleElement28$C$TupleElement29$C$TupleElement30$C$TupleElement31$C$TupleElement32$C$TupleElement33$C$TupleElement34$C$TupleElement35$C$TupleElement36$C$TupleElement37$C$TupleElement38$C$TupleElement39$C$TupleElement40$C$TupleElement41$C$TupleElement42$C$TupleElement43$C$TupleElement44$C$TupleElement45$C$TupleElement46$C$TupleElement47$RP$$u20$as$u20$frame_support..traits..hooks..OnInitialize$LT$BlockNumber$GT$$GT$13on_initialize17h252b62108cdb4ff8E(i32,i32)
	._ZN15frame_executive104Executive$LT$System$C$Block$C$Context$C$UnsignedValidator$C$AllPalletsWithSystem$C$COnRuntimeUpgrade$GT$16initialize_block17h25c01251da9ce46aE(i32)
	._ZN15frame_executive104Executive$LT$System$C$Block$C$Context$C$UnsignedValidator$C$AllPalletsWithSystem$C$COnRuntimeUpgrade$GT$13execute_block17h04bd9202b85e1554E(i32)
	.Core_execute_block(i32,i32) i64	pkg=sync
```

---
**edit**
After a long debug session with @jimjbrettj™ we found that the allocation problem was related with long running instances and that is not trivial to clean up the runtime memory after a block execution (manually clean up), so here is a suggestion of what can be done to solve and a reference to substrate (https://github.com/paritytech/polkadot-sdk/blob/0c6c837f689a287583508506e342ba07687e8d26/substrate/client/executor/wasmtime/src/runtime.rs#L170C39-L170C55)

- [ ] Instantiate a runtime for every block execution

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Detailed error description in issue: https://github.com/ChainSafe/gossamer/issues/3774
- This PR introduces:
  - Refactored the method `(*Instance) Exec` to create a wazero runtime instance with a new allocator to execute the runtime exported method
  - These changes are a fix for two problems `allocator out of space` and `wasm out of bounds memory access`
  - After checking out substrate `runtime/wasmtime` methods we notice they have a strategy called `RecreateInstance` that is responsible for instantiate a new `wasmtime` runtime instance and perform the method call using a fresh new allocator ([substrate call_impl](https://github.com/paritytech/polkadot-sdk/blob/0c6c837f689a287583508506e342ba07687e8d26/substrate/client/executor/wasmtime/src/runtime.rs#L170C39-L170C55) and [strategy](https://github.com/paritytech/polkadot-sdk/blob/0c6c837f689a287583508506e342ba07687e8d26/substrate/client/executor/wasmtime/src/runtime.rs#L143))
  - It is not possible to use a single runtime instance only refreshing the allocator since it leads to `wasm out of bounds memory access`.

## Tests

Before the changes we were facing a panic while executing block #2 on paseo network
image

After the changes we can execute entire chains
image

## Issues

- Closes https://github.com/ChainSafe/gossamer/issues/3774

## Primary Reviewer

@jimjbrettj @timwu20

### CODE REVIEW

File: lib/runtime/wazero/instance.go [hunk 1]
```diff
@@ -458,32 +454,49 @@ var ErrExportFunctionNotFound = errors.New("export function not found")
...
+	defer i.Unlock()
 
+	mod, rt, err := newRuntimeInstance(context.Background(), i.code)
+	if err != nil {
+		return nil, fmt.Errorf("creating runtime instace: %w", err)
+	}
 	defer func() {
-		i.Context.Allocator = nil
-		i.Unlock()
+		err := mod.Close(context.Background())
```
Comment (jimjbrettj) [inline lib/runtime/wazero/instance.go:464]: nit/maybe helper func to close both would be nice? Up to you

**Review by jimjbrettj — APPROVED** [review]
lgtm, just needs to fix tests

File: lib/runtime/wazero/instance_test.go [hunk 2]
```diff
@@ -1096,7 +1095,7 @@ func newTrieFromPairs(t *testing.T, filename string) trie.Trie {
 		entries[pairArr[0].(string)] = pairArr[1].(string)
 	}
 
-	tr, err := inmemory.LoadFromMap(entries, trie.V0)
```
Comment (dimartiro) [inline lib/runtime/wazero/instance_test.go:1099]: thks! ❤️

File: lib/runtime/wazero/instance.go [hunk 3]
```diff
@@ -458,32 +454,49 @@ var ErrExportFunctionNotFound = errors.New("export function not found")
 
 func (i *Instance) Exec(function string, data []byte) (result []byte, err error) {
 	i.Lock()
-	i.Context.Allocator = allocator.NewFreeingBumpHeapAllocator(i.heapBase)
+	defer i.Unlock()
```
Comment (timwu20) [inline lib/runtime/wazero/instance.go:457]: you should probably add this unlock to the defer on line 463 to ensure it unlocks after closing the runtime instance.

File: lib/runtime/wazero/instance_test.go [hunk 4]
```diff
@@ -1096,7 +1095,7 @@ func newTrieFromPairs(t *testing.T, filename string) trie.Trie {
...
+	tr, err := inmemory_trie.LoadFromMap(entries, trie.V0)
```
Comment (timwu20) [inline lib/runtime/wazero/instance_test.go:1098]:
```suggestion
	tr, err := inMemoryTrie.LoadFromMap(entries, trie.V0)
```

File: lib/runtime/wazero/instance.go (hunk 3 above)
Comment (EclesioMeloJunior) [reply lib/runtime/wazero/instance.go:457]: the defers are like a stack of calls, so the last in will be the first out, the last definition will be executed first and so on so forth...

```go
defer execA() // first defer will be executed lastly 
defer execB() // lastly defer will be the first to be executed
```

Comment (timwu20) [reply lib/runtime/wazero/instance.go:457]: TIL

**Comment by EclesioMeloJunior** [comment]
closing in favor of https://github.com/ChainSafe/gossamer/pull/3879

**Pull request closed without merging**